File: packages/vuejs/src/functions/currencyConverter.ts
/**
 * Currency exchange rates interface
 */
export interface ExchangeRates {
  [currency: string]: number;
}

/**
 * Format currency amount
 * @param amount - The amount to format
 * @param currency - Currency code (e.g., 'USD', 'EUR')
 * @param locale - Locale string (e.g., 'en-US', 'id-ID')
 * @returns Formatted currency string
 */
export function formatCurrency(
  amount: number,
  currency: string = 'USD',
  locale: string = 'en-US'
): string {
  try {
    return new Intl.NumberFormat(locale, {
      style: 'currency',
      currency: currency,
    }).format(amount);
  } catch (error) {
    // Fallback formatting
    return `${currency} ${amount.toFixed(2)}`;
  }
}

/**
 * Convert amount between currencies
 * @param amount - The amount to convert
 * @param fromCurrency - Source currency code
 * @param toCurrency - Target currency code
 * @param rates - Exchange rates object
 * @returns Converted amount
 */
export function convertCurrency(
  amount: number,
  fromCurrency: string,
  toCurrency: string,
  rates: ExchangeRates
): number {
  if (fromCurrency === toCurrency) {
    return amount;
  }
  
  // Convert to base currency (USD) first, then to target currency
  const fromRate = rates[fromCurrency] || 1;
  const toRate = rates[toCurrency] || 1;
  
  const inUSD = amount / fromRate;
  return inUSD * toRate;
}

/**
 * Parse currency string to number
 * @param currencyString - String with currency formatting
 * @returns Parsed number
 */
export function parseCurrency(currencyString: string): number {
  // Remove currency symbols and separators, keep decimal point
  const cleaned = currencyString.replace(/[^0-9.-]/g, '');
  return parseFloat(cleaned) || 0;
}

File: packages/flutter/lib/functions/currency_converter.dart
import 'package:intl/intl.dart';

/// Format currency amount
///
/// [amount] The amount to format
/// [currency] Currency code (e.g., 'USD', 'EUR')
/// [locale] Locale string (e.g., 'en_US', 'id_ID')
String formatCurrency(double amount, [String currency = 'USD', String locale = 'en_US']) {
  try {
    final formatter = NumberFormat.currency(
      locale: locale,
      symbol: _getCurrencySymbol(currency),
      decimalDigits: 2,
    );
    return formatter.format(amount);
  } catch (e) {
    // Fallback formatting
    return '$currency ${amount.toStringAsFixed(2)}';
  }
}

/// Convert amount between currencies
///
/// [amount] The amount to convert
/// [fromCurrency] Source currency code
/// [toCurrency] Target currency code
/// [rates] Exchange rates map
double convertCurrency(
  double amount,
  String fromCurrency,
  String toCurrency,
  Map<String, double> rates,
) {
  if (fromCurrency == toCurrency) {
    return amount;
  }
  
  // Convert to base currency (USD) first, then to target currency
  final fromRate = rates[fromCurrency] ?? 1.0;
  final toRate = rates[toCurrency] ?? 1.0;
  
  final inUSD = amount / fromRate;
  return inUSD * toRate;
}

/// Parse currency string to number
///
/// [currencyString] String with currency formatting
double parseCurrency(String currencyString) {
  // Remove currency symbols and separators, keep decimal point
  final cleaned = currencyString.replaceAll(RegExp(r'[^0-9.-]'), '');
  return double.tryParse(cleaned) ?? 0.0;
}

/// Get currency symbol for currency code
String _getCurrencySymbol(String currency) {
  const symbols = {
    'USD': '\$',
    'EUR': '€',
    'GBP': '£',
    'JPY': '¥',
    'IDR': 'Rupiah',
    'SGD': 'S\$',
    'MYR': 'RM',
  };
  return symbols[currency] ?? currency;
}
